src/java.base/share/classes/jdk/crac/RCULock.java

```diff
@@ -0,0 +1,255 @@
+    private SwitchPoint unlockSwitchPoint = new SwitchPoint();
+    private volatile MethodHandle lockImpl;
+    private volatile MethodHandle unlockImpl;
+
+    static {
+        initFieldOffsets();
+    }
+
+    private static native void initFieldOffsets();
+
+    /**
+     * Inspect the classes for all methods (including private ones) marked with
+     * the {@link RCULock.Critical} annotation and create a lock that can protect
+     * these methods with read lock.
+     *
+     * @param classes List of classes to be inspected.
+     * @return New lock instance.
+     */
+    public static RCULock forClasses(Class<?>... classes) {
+        ArrayList<Method> methods = new ArrayList<>();
+        for (Class<?> cls : classes) {
+            while (cls != null && cls != Object.class) {
+                for (Method m : cls.getDeclaredMethods()) {
+                    if (m.isAnnotationPresent(Critical.class)) {
+                        methods.add(m);
+                    }
+                }
+                cls = cls.getSuperclass();
+            }
+        }
+        return new RCULock(methods);
+    }
+
+    /**
+     * Create a lock that can protect selected methods with read-locking.
+     *
+     * @param readCriticalMethods List of methods with the read-critical section.
+     */
+    public RCULock(Iterable<Method> readCriticalMethods) {
+        this(StreamSupport.stream(readCriticalMethods.spliterator(), false)
+                .map(m -> m.getDeclaringClass().getName() + "." + m.getName() + "(" +
+                Arrays.stream(m.getParameterTypes())
+                        .map(Class::descriptorString).collect(Collectors.joining())
+                + ")" + m.getReturnType().descriptorString()).toArray(String[]::new));
+    }
+
+    /**
+     * Create a lock that can protect selected methods with read-locking.
+     * <p>
+     * The signatures follow the form
+     * <pre>my.package.MyClass.foobar(another/package/SomeClass;Z)V</pre>
+     *
+     * @param readCriticalMethods List of signatures for methods invoked in the read-critical section.
+     */
+    public RCULock(String[] readCriticalMethods) {
+        initSwitchPoints();
+        //noinspection CapturingCleaner
+        CleanerFactory.cleaner().register(this, this::destroy);
+        Arrays.sort(readCriticalMethods);
+        init(readCriticalMethods);
+    }
+
+    private native void init(String[] readCriticalMethods);
+
+    private native void destroy();
+
+    private void initSwitchPoints() {
+        try {
+            MethodType voidType = MethodType.methodType(void.class);
+            MethodHandle noop = MethodHandles.lookup().findSpecial(RCULock.class, "noop", voidType, RCULock.class);
+            MethodHandle readLockImpl = MethodHandles.lookup().findSpecial(RCULock.class, "readLockImpl", voidType, RCULock.class);
+            MethodHandle readUnlockImpl = MethodHandles.lookup().findSpecial(RCULock.class, "readUnlockImpl", voidType, RCULock.class);
```

Comment:
Creating a Lookup object is expensive.  Better to cache in a local
```suggestion
            MethodHandles.Lookup lookup = MethodHandles.lookup();
            MethodHandle noop = lookup.findSpecial(RCULock.class, "noop", voidType, RCULock.class);
            MethodHandle readLockImpl = lookup.findSpecial(RCULock.class, "readLockImpl", voidType, RCULock.class);
            MethodHandle readUnlockImpl = lookup.findSpecial(RCULock.class, "readUnlockImpl", voidType, RCULock.class);
```